Frameworks/CoreGraphics/CGGradient.mm

```diff
@@ -18,186 +18,163 @@
 }
 
 /**
  @Status Interoperable
 */
-CGGradientRef CGGradientCreateWithColors(CGColorSpaceRef colorSpace, CFArrayRef colors, const float* locations) {
-    CGGradientRef ret = new __CGGradient();
-    ret->initWithColors(colors, locations, colorSpace);
-
-    return ret;
+void CGGradientRelease(CGGradientRef gradient) {
+    RETURN_IF(!gradient);
+    CFRelease(static_cast<CFTypeRef>(gradient));
 }
 
 /**
  @Status Interoperable
 */
-void CGGradientRelease(CGGradientRef gradient) {
-    CFRelease((id)gradient);
+CGGradientRef CGGradientRetain(CGGradientRef gradient) {
+    RETURN_NULL_IF(!gradient);
+    CFRetain(static_cast<CFTypeRef>(gradient));
+    return gradient;
+}
+
+static inline bool __isValidGradientColorspaceModel(CGColorSpaceRef colorspace) {
+    RETURN_RESULT_IF_NULL(colorspace, false);
+
+    // The colorspace cannot be kCGColorSpaceModelPattern || kCGColorSpaceModelIndexed
+    CGColorSpaceModel colorSpaceModel = CGColorSpaceGetModel(colorspace);
+    if ((colorSpaceModel == kCGColorSpaceModelPattern) || (colorSpaceModel == kCGColorSpaceModelIndexed)) {
+        TraceError(TAG, L"Invalid colorspace %ld for gradient", colorSpaceModel);
+        return false;
+    }
+
+    return true;
 }
 
 /**
  @Status Interoperable
 */
-CGGradientRef CGGradientRetain(CGGradientRef gradient) {
-    CFRetain((id)gradient);
-    return gradient;
+CGGradientRef CGGradientCreateWithColorComponents(CGColorSpaceRef colorSpace,
+                                                  const CGFloat* components,
+                                                  const CGFloat* locations,
+                                                  size_t count) {
+    RETURN_NULL_IF(!colorSpace);
+    RETURN_NULL_IF(!components);
+    RETURN_NULL_IF(count == 0);
+    RETURN_NULL_IF(!__isValidGradientColorspaceModel(colorSpace));
+    // location can be null
+    return __CGGradient::CreateInstance(kCFAllocatorDefault, colorSpace, components, locations, count);
 }
 
 /**
- @Status Stub
- @Notes
+ @Status Interoperable
 */
-CFTypeID CGGradientGetTypeID() {
-    UNIMPLEMENTED();
-    return StubReturn();
+CGGradientRef CGGradientCreateWithColors(CGColorSpaceRef colorSpace, CFArrayRef colors, const CGFloat* locations) {
+    RETURN_NULL_IF(!colorSpace);
+    RETURN_NULL_IF(!colors);
+    RETURN_NULL_IF(!__isValidGradientColorspaceModel(colorSpace));
+    // location can be null
+    return __CGGradient::CreateInstance(kCFAllocatorDefault, colorSpace, colors, locations);
 }
 
-CGFloat* _CGGradientGetStopLocation(CGGradientRef gradient) {
-    RETURN_NULL_IF(!gradient);
-    return gradient->_locations;
+std::vector<CGFloat> _CGGradientGetStopLocations(CGGradientRef gradient) {
+    RETURN_RESULT_IF_NULL(gradient, {});
+    return gradient->GetStopLocations();
 }
 
-CGFloat* _CGGradientGetColorComponents(CGGradientRef gradient) {
-    RETURN_NULL_IF(!gradient);
-    return gradient->_components;
+std::vector<CGFloat> _CGGradientGetColorComponents(CGGradientRef gradient) {
```

Comment:
since the consumers do some manipulations (e.g append).
It would be best to return as this for now.